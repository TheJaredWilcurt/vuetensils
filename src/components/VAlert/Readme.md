### Styled Examples

```vue
<template>
  <v-alert class="error">
    <h3>Danger!</h3>
    <p>Something went horribly wrong!!!</p>
  </v-alert>
</template>

<script>
export default {};
</script>

<style>
.vts-alert.error {
  border: 1px solid currentColor;
  border-radius: 4px;
  padding: 0 10px;
  color: #900;
  background: #fdd;
}
</style>
```

```vue
<template>
  <div class="styled">
    <v-alert class="info" dismissible>
      Did you know that giraffes are afraid of heights?
    </v-alert>
  </div>
</template>

<script>
export default {};
</script>

<style>
.vts-alert.info {
  display: flex;
  justify-content: space-between;
  align-items: center;
  border: 1px solid currentColor;
  border-radius: 4px;
  padding: 10px;
  color: #009;
  background: #dff;
}

.vts-alert.info .vts-alert__dismiss {
  border: 0;
  font: inherit;
  background: transparent;
}
</style>
```

### Unstyled Examples

```vue
<template>
  <v-alert>
    Here's something I need to tell you about.
  </v-alert>
</template>

<script>
export default {};
</script>
```

### Using v-model

With `v-model`, you can toggle an alert on or off.

```vue
<template>
  <div>
    <button @click="alert = !alert">Toggle alert</button>

    <v-alert v-model="alert">
      I can be toggled
    </v-alert>
  </div>
</template>

<script>
export default {
  data: () => ({
    alert: false,
  }),
};
</script>
```

### Dismissible

Allow users to hide an alert

```vue
<template>
  <v-alert dismissible>
    Click that button to make me go away
  </v-alert>
</template>

<script>
export default {};
</script>
```

```vue
<template>
  <div>
    <button @click="alert = true">Show alert</button>

    <v-alert v-model="alert" dismissible>
      Click that button to make me go away
    </v-alert>
  </div>
</template>

<script>
export default {
  data: () => ({
    alert: false,
  }),
};
</script>
```

You can provide your own content for the dismiss button with the `dismiss` slot.

```vue
<template>
  <v-alert dismissible>
    Click that button to make me go away

    <template slot="dismiss">
      Hide
    </template>
  </v-alert>
</template>

<script>
export default {};
</script>
```

### Countdown

It also supports providing seconds to the `visible` prop (or `v-model`) so the alert will hide after a short time.

```vue
<template>
  <div>
    <button @click="countdown = 10">Reset time</button>
    <v-alert v-model="countdown" dismissible>
      This alert will dismiss in {{ countdown }} seconds.
    </v-alert>
  </div>
</template>

<script>
export default {
  data: () => ({
    countdown: 10,
  }),
};
</script>
```

### Transitions

```vue
<template>
  <div>
    <button @click="alert = !alert">Toggle alert</button>

    <v-alert v-model="alert" transition="fade">
      I can be toggled
    </v-alert>
  </div>
</template>

<script>
export default {
  data: () => ({
    alert: false,
  }),
};
</script>

<style>
.fade-enter-active,
.fade-leave-active {
  transition: 0.5s ease opacity;
}

.fade-enter,
.fade-leave-to {
  opacity: 0;
}
</style>
```

### Custom Classes

This component can accept a `classes` prop to customize the output HTML classes:

```
:classes="{ root: 'root-class', dismiss: 'dismiss-class' }"
```
