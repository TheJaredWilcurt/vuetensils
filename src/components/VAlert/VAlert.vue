<template>
  <transition :name="transition">
    <component
      :is="tag"
      v-if="!dismissed && !!visible"
      role="alert"
      :class="['vts-alert', classes.root]"
    >
      <!-- @slot The default slot content that is shown to the user -->
      <slot />

      <button
        v-if="dismissible"
        :aria-label="dismissLabel"
        :class="['vts-alert__dismiss', classes.dismiss]"
        @click="dismiss"
      >
        <!-- @slot The dismiss button content -->
        <slot name="dismiss">
          &times;
        </slot>
      </button>
    </component>
  </transition>
</template>

<script>
/**
 * NOTES Concerning :visible prop. 
 * Per W3C specifications:
 * https://www.w3.org/TR/wai-aria-practices/#alert
 * --
 * "It is also important to avoid designing alerts that disappear automatically. 
 * An alert that disappears too quickly can lead to failure to meet WCAG 2.0 success criterion 2.2.3. 
 * Another critical design consideration is the frequency of interruption caused by alerts. 
 * Frequent interruptions inhibit usability for people with visual and cognitive disabilities, 
 * which makes meeting the requirements of WCAG 2.0 success criterion 2.2.4 more difficult."
 */

/**
 * A simple component for notifiying users of specific information. Good for informative snippets, error messages, and more. It can be shown or hidden dynamically, and even supports auto-hiding after a given time.
 */
export default {
  name: 'VAlert',
  model: {
    prop: 'visible',
    event: 'update',
  },

  props: {
    /**
     * HTML tag used to wrap the component.
     */
    tag: {
      type: String,
      default: 'div',
    },
    /**
     * Determines whether the alert is visible. Also binds with `v-model`.
     */
    visible: {
      type: [Boolean, Number],
      default: true,
    },
    /**
     * Allows a user to dismiss this alert.
     */
    dismissible: Boolean,
    /**
     * Aria-label that is not visibly, but screen readers will read for the dismiss button.
     */
    dismissLabel: {
      type: [String, Boolean],
      default: 'Dismiss this alert',
    },
    /**
     * The transition name if you want to add one.
     */
    transition: {
      type: String,
      default: undefined,
    },

    classes: {
      type: Object,
      default: () => ({}),
    },
  },

  data: () => ({
    dismissed: false,
    timerId: null,
  }),

  watch: {
    visible: {
      handler(visible) {
        if (visible) {
          this.dismissed = false;
        }
        if (typeof visible === 'number') {
          this.clearTimer(); // Clear timers in case this.visible watcher adds multiples
          this.countdown();
        }
      },
      immediate: true,
    },
  },

  beforeDestroy() {
    this.clearTimer();
  },

  methods: {
    dismiss() {
      /**
       * Fired when a user manually dismissed an alert
       *
       * @event dismiss
       * @type { undefined }
       */
      this.$emit('dismiss');
      this.dismissed = true;
      if (typeof this.visible === 'number') {
        this.$emit('update', 0);
        this.clearTimer();
      } else {
        this.$emit('update', false);
      }
    },

    countdown() {
      const { visible } = this;
      if (visible <= 0) return;

      this.timerId = setTimeout(() => {
        /**
         * Fired whenever the visibility changes. Either through user interaction, or a countdown timer
         *
         * @event update
         * @type { boolean|number }
         */
        this.$emit('update', visible - 1);
      }, 1000);
    },

    clearTimer() {
      const { timerId } = this;
      if (timerId) {
        clearInterval(timerId);
        this.timerId = null;
      }
    },
  },
};
</script>
