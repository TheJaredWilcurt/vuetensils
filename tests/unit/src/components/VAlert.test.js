import { render } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import VAlert from '@components/VAlert/VAlert.vue';

describe('VAlert', () => {
  describe('Default Functionality', () => {

    it('render correctly without props', () => {
      const { getByRole } = render(VAlert);
      const alertBox = getByRole('alert');
      expect(alertBox).toMatchSnapshot();
    });

    describe('Dismiss & Timing visibility', () => {

      it('render default dismiss button', () => {
        const { getByLabelText } = render(VAlert, {
          props: {
            dismissible: true,
          },
        });
  
        const dismissButton = getByLabelText('Dismiss this alert');
        expect(dismissButton).toBeVisible();
      });

      it('dismisses @default slot content', () => {
        const { getByText, getByLabelText } = render(VAlert, {
          props: {
            dismissible: true,
          },
          slots: {
            default: `<span>Testing Content in @default slot</span>`
          }
        });
  
        const alertBox = getByText('Testing Content in @default slot');
        const dismissButton = getByLabelText('Dismiss this alert');
        
        expect(alertBox).toBeVisible();
        userEvent.click(dismissButton);        
        expect(alertBox).not.toBeVisible();
      });
    });
  });
});
