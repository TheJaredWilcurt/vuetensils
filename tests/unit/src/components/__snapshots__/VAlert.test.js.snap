// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VAlert Default Functionality render correctly without props 1`] = `
<div
  class="vts-alert"
  role="alert"
>
   
  <!---->
</div>
`;
